src/topology/continuous_function/stone_weierstrass.lean

```diff
@@ -0,0 +1,353 @@
+    { rw [show U x y = (f - g x y : C(X, ℝ)) ⁻¹' set.Iio ε, { ext z, simp [sub_lt], }],
+      exact is_open.preimage (by continuity) is_open_Iio, },
+    { rw [set.mem_set_of_eq, w₂],
+      exact sub_lt_self _ pos, }, },
+
+  -- Fixing `x` for a moment, we have a family of functions `λ y, g x y`
+  -- which on different patches (the `U x y`) are greater than `f z - ε`.
+  -- Taking the supremum of these functions
+  -- indexed by a finite collection of patches which cover `X`
+  -- will give us an element of `A` that is globally greater than `f z - ε`.
+
+  -- Since `X` is compact, for every `x` there is some finset `ys t`
+  -- so the union of the `U x y` for `y ∈ ys x` still covers everything.
+  let ys : Π x, finset X := λ x, (compact_space.elim_nhds_subcover (U x) (U_nhd_y x)).some,
+  let ys_w : ∀ x, (⋃ y ∈ ys x, U x y) = ⊤ :=
+    λ x, (compact_space.elim_nhds_subcover (U x) (U_nhd_y x)).some_spec,
+  have ys_nonempty : ∀ x, (ys x).nonempty :=
+    λ x, set.nonempty_of_union_eq_top_of_nonempty _ _ nX (ys_w x),
+
+  -- Thus for each `x` we have the desired `h x : A` so `f z - ε < h x z` everywhere.
+  let h : Π x, L := λ x,
+    ⟨(ys x).sup' (ys_nonempty x) (λ y, (g x y : C(X, ℝ))),
+      finset.sup'_mem _ sup_mem _ _ _ (λ y _, (g x y).2)⟩,
+  have lt_h : ∀ x z, f z - ε < h x z,
+  { intros x z,
+    obtain ⟨y, ym, zm⟩ := set.exists_set_mem_of_union_eq_top _ _ (ys_w x) z,
+    dsimp [h],
+    simp only [finset.lt_sup'_iff, continuous_map.sup'_apply],
+    exact ⟨y, ym, zm⟩, },
+
+  -- For each `x y`, we define `V x y` to be `{ z | g x y z < f z + ε }`,
+  -- and observe this is a neighbourhood of `x`.
+  let V : Π x y, set X := λ x y, { z | g x y z < f z + ε },
+  have V_nhd_x : ∀ x y, V x y ∈ 𝓝 x,
+  { intros x y,
+    refine mem_nhds_sets _ _,
+    { rw [show V x y = (g x y - f : C(X, ℝ)) ⁻¹' set.Iio ε, { ext z, simp [sub_lt_iff_lt_add'], }],
+      exact is_open.preimage (by continuity) is_open_Iio, },
+    { rw [set.mem_set_of_eq, w₁],
+      exact (lt_add_iff_pos_right _).mpr pos, }, },
+
+  -- For each `x`, we can take the finite intersection of the `V x y` corresponding to `y ∈ ys x`.
+  let W : Π x, set X := λ x, (⋂ y ∈ ys x, V x y),
+  -- This is still a neighbourhood of `x`.
+  have W_nhd : ∀ x, W x ∈ 𝓝 x :=
+    λ x, (filter.bInter_finset_mem_sets (ys x)).2 (λ y m, V_nhd_x x y),
+  -- Locally on each `W x`, we have `h x z < f z + ε`, since `h x` is a supremum of the `g x y`.
+  have h_lt : ∀ (x) (z ∈ W x), h x z < f z + ε,
+  { intros x z zm,
+    dsimp [h],
+    simp only [continuous_map.sup'_apply, finset.sup'_lt_iff],
+    intros y ym,
+    convert set.mem_of_mem_of_subset zm (set.bInter_subset_of_mem ym), },
+
+  -- Since `X` is compact, there is some finset `ys t`
+  -- so the union of the `W x` for `x ∈ xs` still covers everything.
+  let xs : finset X := (compact_space.elim_nhds_subcover W W_nhd).some,
+  let xs_w : (⋃ x ∈ xs, W x) = ⊤ :=
+    (compact_space.elim_nhds_subcover W W_nhd).some_spec,
```

Comment:
can you use the `choose` version again?

Reply:
As above?

Reply:
If you don't like it above, I guess you won't like it here either :-)